core/migration.go

```diff
@@ -150,6 +162,43 @@ func (m *Migration) Run(context *Context) {
 	}
 }
 
+func (m *Migration) addSubIds(context *Context, task *AddSubIds) {
+	users := context.GetUsersInGroup(task.GroupName)
+	users = append(users, context.GetRootUser())
+
+	if _, err := os.Stat("/etc/subuid"); err != nil {
+		if _, err = os.Create("/etc/subuid"); err != nil {
+			waterlog.Warnf("\tUnable to create subuid file due to error: %s\n", err)
+		} else {
+			for _, user := range users {
+				if err = context.AddSubUids(&user, task.RangeStart, task.RangeEnd); err != nil {
+					waterlog.Warnf("\tFailed to add subuids to user %s due to error: %s\n", user.Name, err)
+				} else {
+					waterlog.Debugf("\tSuccessfully added subuids to user %s\n", user.Name)
+				}
+			}
+		}
+	} else {
+		waterlog.Debugf("\t/etc/subuid already exists, skipping\n")
+	}
+
+	if _, err := os.Stat("/etc/subgid"); err != nil {
+		if _, err = os.Create("/etc/subgid"); err != nil {
+			waterlog.Warnf("\tUnable to create subuid file due to error: %s\n", err)
+		} else {
+			for _, user := range users {
+				if err = context.AddSubGids(&user, task.RangeStart, task.RangeEnd); err != nil {
+					waterlog.Warnf("\tFailed to add subuids to user %s due to error: %s\n", user.Name, err)
+				} else {
+					waterlog.Debugf("\tSuccessfully added subuids to user %s\n", user.Name)
+				}
+			}
+		}
+	} else {
+		waterlog.Debugf("\t/etc/subgid already exists, skipping")
+	}
+}
```

Comment:
This really needs to be split for readability and reduced nesting. Also: the file created by `os.Create` is not closed.

For example:

```suggestion
func (m *Migration) addSubIds(context *Context, task *AddSubIds) {
	users := context.GetUsersInGroup(task.GroupName)
	users = append(users, context.GetRootUser())

	m.createSubUidFile(context, task, users)
	m.createSubGidFile(context, task, users)
}

func createFileIfNotExists(path string) (created bool) {
	if _, err := os.Stat(path); err == nil {
		waterlog.Debugf("\t%s already exists, skipping", path)

		return false
	}

	fh, err := os.Create(path)
	if err != nil {
		waterlog.Warnf("\tUnable to create %s due to error: %s\n", path, err)

		return false
	}

	fh.Close()

	return true
}

func (m *Migration) createSubUidFile(context *Context, task *AddSubIds, users []User) {
	if !createFileIfNotExists("/etc/subuid") {
		return
	}

	for _, user := range users {
		if err := context.AddSubUids(&user, task.RangeStart, task.RangeEnd); err != nil {
			waterlog.Warnf("\tFailed to add subuids to user %s due to error: %s\n", user.Name, err)
			return
		}

		waterlog.Debugf("\tSuccessfully added subuids to user %s\n", user.Name)
	}
}

func (m *Migration) createSubGidFile(context *Context, task *AddSubIds, users []User) {
	if !createFileIfNotExists("/etc/subgid") {
		return
	}

	for _, user := range users {
		if err := context.AddSubGids(&user, task.RangeStart, task.RangeEnd); err != nil {
			waterlog.Warnf("\tFailed to add subgids to user %s due to error: %s\n", user.Name, err)
			return
		}

		waterlog.Debugf("\tSuccessfully added subgids to user %s\n", user.Name)
	}
}
```